index.bs

```diff
@@ -738,17 +738,91 @@ that should be considered in the development of new features, notably:
+
+Whenever these differences could be observable to web content
+and there isn't a good reason to allow variation between implementations,
+it is important that they be specified unambiguously,
+so that implementations interoperate
+and web content doesn't end up
+depending on the behavior of one implementation and failing in another
+(which can cause higher development costs to fix the content for all browsers,
+or costs to users of broken content if it wasn't detected).
+
+However, sequences of steps are not the only way of making the specification precise enough.
+When there are particular characteristics
+that specification authors want to ensure the specification always satisfies,
+it may be better to use a less algorithmic way of formal specification
+that ensures these characteristics are always true.
+Developing these other ways of specifying behavior formally
+is more important when the algorithms would be more repetitive.
+These formal specification techniques can include dependencies on
+languages such as [[WEBIDL]], on grammar productions,
+or the development of new languages for defining pieces of the specification.
+Merely having a formal syntax is not sufficient, though;
+the requirements that the formal definitions place on implementations must still be clearly defined,
+which generally requires specifying some algorithms (or, sometimes, another layer of formal syntax) that define what the formal syntax means,
+and then using the more-readable formal syntax to avoid
+excessive repetition of the less-readable algorithms.
+There are also possibilities in-between, for example,
+if it is desirable to ensure that an algorithm can be implemented with a state machine,
+it may be preferable to specify it in an algorithmic way that incorporates explicit states
+so that the specification authors avoid accidentally creating more states than intended.
+
+When specifying in an algorithmic way,
+it is generally good to match the algorithms that implementations would use,
+even if this is somewhat less convenient for specification authors.
+For example, it is better to specify CSS selector matching
+by moving right to left across <a>combinators</a>,
+since that is what implementations do.
+This avoids the risk of future problems caused by new features
+that would be slightly different (in unimportant ways)
+between implementations based on one algorithm and implementations based on a different one.
+Thus, it avoids small future divergence in behavior between specification and implementations.
+It also helps to keep the costs of specifying features aligned with implementing them,
+since the complexity of adding a feature to the specification is more likely to be related
+to the cost of adding it to implementations.
+
+Algorithms in specifications should also reflect best practices in programming.
+For example, they should explictly describe the inputs and outputs of the algorithm,
```

Comment:
I don't think we typically specify outputs upfront today, but they're clearish from return statements. Perhaps something to reconsider though.

Reply:
I strongly prefer specifying outputs, personally. I'm definitely not *strict* about specifying them myself, but algorithms like <https://drafts.csswg.org/css-syntax/#consume-ident-like-token> make it clear what the range of the return type is when it may not be obvious from context, so you don't have to track variables and returns.

Reply:
When I wrote about specifying outputs, I was thinking more about having other bits of algorithm that peek inside the algorithm and use something written as internal to it as an additional output (i.e., the opposite of the stack introspection issue).  But I think Tab's idea about being clear about result types is good as well.

I guess I'm not sure if the text I have already is fine for covering that, or if there's a need to go into detail about the distinction...

Reply:
I think this is probably fine.